### ISSUE TITLE
A way to specify externals and versions and have them downloaded

### ISSUE THREAD

**Issue by chr15m** [pure-data/deken#176]
Related: #92 #93 #94

Use-case: a person distributing a patch wants the end user of the patch to be able to quickly and easily get the right set of external dependencies to run the patch.

Design sketch:

 * New menu item under "Find externals" -> "Install dependencies".
 * Menu item appears if current patch (window) has a file `externals.txt` in the folder where it is saved.
 * Format of the `externals.txt` file copies prior art: Javascript's `package.json`, Python's `requirements.txt`, and Ruby's `Gemfile`.
 * When the menu item is selected the deken plugin searches for and tries to install the first search hit of each line in the file. [search would need to be amended to allow for version-targeted searching]
 * By default installs to the folder where the patch itself resides with the option to install to the Pd-wide default. This is so the end user can install the dependencies of a patch without messing up their system externals.
 * Example of `externals.txt` file format:

```
freeverb~ v1.2.3
creb
zexy v2.2.6
```

**Comment by umlaeute** [comment]
one problem: the user already has `zexy v2.3` installed. deken has no notion of the current state, so it might
- downgrade the library
- install multiple different versions of the same library

**Comment by chr15m** [comment]
Yep that's a tough problem. I think if this mechanism confines itself to the patch-local folder only then it could avoid the issue for the most part.

When I think about the situations where somebody would use this functionality patch-local-folder install seems like the best default. For the most part I expect it would be: "I've downloaded this patch from the internet and it has a bunch of dependencies. I want to try it out but I don't want to mess up my carefully curated Pd externals collection."

So it would behave the same as `node_modules` or Python `virtualenv` in that respect.

**Comment by umlaeute** [comment]
btw, this is implemented in the cmdline tool:

~~~
$ cat <<EOF >requirements.txt
zexy==2.2.5
iemguts>=0.3
creb
EOF
$ deken -v install --no-verify -r requirements.txt --installdir .
~~

- there's currently a bug that will prevent the installation of unversioned libraries (`creb`) if there are versioned libraries present (`zexy==2.2.5`, `iemguts>=0.3`)
- this issue is (of course) mainly about getting this functionality in the GUI-plugin!

**Comment by chr15m** [comment]
Wow that looks cool!

**Comment by umlaeute** [comment]
and it's been there since about ba853a329dbb850c133c9c14ae58da98bb2b9b4c ...

**Comment by mxa** [comment]
Instead of an `externals.txt` it could also just parse the pd patch for `[declare -lib/-path]`. Example: You open `MAIN.pd` and (there are errors 'could not create object' when opening), then deken would ask if you want to install the declared libs freeverb, creb and zexy and reload the patch.

**Comment by umlaeute** [comment]
the plan is slightly different:
- prerequisite: you can search for libraries and/or objects
- clicking on the "could not create" error-message in the Pd-console, will open up the deken search dialog, with the missing object name already filled in (and the proper search mode selected)

the above part is pretty easy (and is already present in pure-data's [`feature/dekenception` branch](https://github.com/pure-data/pure-data/tree/feature/dekenception)).

the problems start with extracting *missing* libraries from `[declare]` (as the `[declare]` object itself doesn't do any loading in the general case).
also, a binary that cannot be loaded by `[declare -lib]` does not necessarily mean that all things are broken anyhow (see https://github.com/pure-data/pure-data/issues/281).

**Comment by umlaeute** [comment]
i would not want to (re)install all required libraries with a simple click: currently re-installing a deken package means, that a folder is wiped from the harddisk, which can have very undesired side-effects.

**Comment by umlaeute** [comment]
i've finally opened a PR against Pd to enable this https://github.com/pure-data/pure-data/pull/1547
